### PR TITLE
Functionality to add IPTV/URL via Bouquet Editor

### PR DESCRIPTION
**Pull request by wedebe**
Functionality to add IPTV/URL via Bouquet Editor

Known issues:
- errors not presented to user (eg. service already exists in bouquet)

### CODE REVIEW

File: plugin/controllers/BouquetEditor.py [hunk 1]
```diff
@@ -304,7 +303,7 @@ def addServiceToBouquet(self, param):
 		mutableBouquetList = self.getMutableList(bouquetRef)
 		if mutableBouquetList is not None:
 			if sRefUrl:
-				ref = eServiceReference(4097, 0, sRef + ':' + sName)
+				ref = eServiceReference(4097, 0, sRef)
```
Comment (jbleyel) [inline plugin/controllers/BouquetEditor.py:306]: Please don't do this.

File: plugin/controllers/views/responsive/ajax/channels.tmpl [hunk 2]
```diff
@@ -53,7 +53,7 @@
 
 		#if '4097:' in $channel.ref
 			#set $rname = ":" + $channel.name
-			#set $lnk = $channel.ref.replace("4097:0:1:0:0:0:0:0:0:0:", "")
+			#set $lnk = $channel.ref.replace("4097:0:0:0:0:0:0:0:0:0:", "").replace("4097:0:1:0:0:0:0:0:0:0:", "")
```
Comment (jbleyel) [inline plugin/controllers/views/responsive/ajax/channels.tmpl:56]: Why 4097:0:0:0:0:0:0:0:0:0:?

File: sourcefiles/modern/js/vti-responsive-bqe.js [hunk 3]
```diff
@@ -663,6 +666,53 @@
 				self._addMarker(true);
 			},
 
+			// Callback function for right pane add marker button
+			// Prompts for marker name, marker will be added before selected service
+			_addUrl: function () {
```
Comment (jbleyel) [inline sourcefiles/modern/js/vti-responsive-bqe.js:671]: Wrong comment.

**Review by jbleyel — COMMENTED** [review]
Please check

File: plugin/controllers/BouquetEditor.py (hunk 1 above)
Comment (wedebe) [reply plugin/controllers/BouquetEditor.py:306]: Service name is set further down
```
				# ref = eServiceReference(4097, 0, sRef)
				ref = eServiceReference(4097, 0, sRef + ':' + sName)
			else:
				ref = eServiceReference(sRef)
			if sName:
				ref.setName(sName)
```

`sRef + ':' + sName` results in a userbouquet entry of
```
#SERVICE 4097:0:0:0:0:0:0:0:0:0:https%3a//www.radioking.com/play/chocolat-radio%3aChocolat Radio:Chocolat Radio
#DESCRIPTION Chocolat Radio
```
when it should be
```
#SERVICE 4097:0:0:0:0:0:0:0:0:0:https%3a//www.radioking.com/play/chocolat-radio:Chocolat Radio
#DESCRIPTION Chocolat Radio
```

File: plugin/controllers/views/responsive/ajax/channels.tmpl (hunk 2 above)
Comment (wedebe) [reply plugin/controllers/views/responsive/ajax/channels.tmpl:56]: This seems to be what's saved to the userbouquet by `addServiceToBouquet`.
I'm not experienced with the numbering so added the all-zeroes as a workaround.

http://192.168.43.147/bouqueteditor/api/addservicetobouquet?sBouquetRef=1%3A7%3A1%3A0%3A0%3A0%3A0%3A0%3A0%3A0%3AFROM+BOUQUET+%22userbouquet.favourites.tv%22+ORDER+BY+bouquet&Name=Choc+Rad&sRefBefore=1%3A0%3A19%3A1B1C%3A802%3A2%3A11A0000%3A0%3A0%3A0%3A&sRefUrl=https%3A%2F%2Fwww.radioking.com%2Fplay%2Fchocolat-radio&_=

Comment (jbleyel) [reply plugin/controllers/views/responsive/ajax/channels.tmpl:56]: My fault.

File: plugin/controllers/BouquetEditor.py (hunk 1 above)
Comment (jbleyel) [reply plugin/controllers/BouquetEditor.py:306]: The result can be:

```
#SERVICE 4097:0:0:0:0:0:0:0:0:0:https%3a//www.radioking.com/play/chocolat-radio:Chocolat Radio
#DESCRIPTION Chocolat Radio
```
or 
```
#SERVICE 4097:0:0:0:0:0:0:0:0:0:https%3a//www.radioking.com/play/chocolat-radio
#DESCRIPTION Chocolat Radio
```

Comment (jbleyel) [reply plugin/controllers/BouquetEditor.py:306]: This is wrong:
```

#SERVICE 4097:0:0:0:0:0:0:0:0:0:https%3a//www.radioking.com/play/chocolat-radio%3aChocolat Radio:Chocolat Radio
#DESCRIPTION Chocolat Radio
```

Comment (wedebe) [reply plugin/controllers/BouquetEditor.py:306]: `%3aChocolat Radio:Chocolat Radio` was caused by `sRef + ':' + sName`, which is why I removed the ` + ':' + sName` part

File: sourcefiles/modern/js/vti-responsive-bqe.js (hunk 3 above)
Comment (wedebe) [reply sourcefiles/modern/js/vti-responsive-bqe.js:671]: Fixed

**Comment by wedebe** [comment]
@jbleyel , could I ask you to investigate why `4097:0:0:0:0:0:0:0:0:0:` is saved (and perhaps fix), please?

**Comment by jbleyel** [comment]
The default IPTV srefs are:
4097:0:1:0:0:0:0:0:0:0: -> TV
4097:0:2:0:0:0:0:0:0:0: -> Radio

/*
 * UNUSED variable from service reference is now used as buffer flag for gstreamer
 * REFTYPE:FLAGS:STYPE:SID:TSID:ONID:NS:PARENT_SID:PARENT_TSID:UNUSED
 *   D  D X X X X X X X X
 * 4097:0:1:0:0:0:0:0:0:0:URL:NAME (no buffering)
 * 4097:0:1:0:0:0:0:0:0:1:URL:NAME (buffering enabled)
 * 4097:0:1:0:0:0:0:0:0:3:URL:NAME (progressive download and buffering enabled)
 *
 * Progressive download requires buffering enabled, so it's mandatory to use flag 3 not 2
 */

4097 can also be 1.

**Comment by wedebe** [comment]
> The default IPTV srefs are:
> 4097:0:1:0:0:0:0:0:0:0: -> TV
> 4097:0:2:0:0:0:0:0:0:0: -> Radio
> 
> /*
> 
> * UNUSED variable from service reference is now used as buffer flag for gstreamer
> * REFTYPE:FLAGS:STYPE:SID:TSID:ONID:NS:PARENT_SID:PARENT_TSID:UNUSED
> * D  D X X X X X X X X
> * 4097:0:1:0:0:0:0:0:0:0:URL:NAME (no buffering)
> * 4097:0:1:0:0:0:0:0:0:1:URL:NAME (buffering enabled)
> * 4097:0:1:0:0:0:0:0:0:3:URL:NAME (progressive download and buffering enabled)
> * Progressive download requires buffering enabled, so it's mandatory to use flag 3 not 2
>   */
> 
> 4097 can also be 1.

Great, thanks!

**Comment by wedebe** [comment]
I'm not sure how to set `stype` along with a url...

`openatv/lib/service/iservice.h`
only seems to handle
`eServiceReference(int type, int flags)',
`eServiceReference(int type, int flags, int data0)` (with increments up to `data4`), 
or
`eServiceReference(int type, int flags, const std::string &path)`

**Pull request merged**